lib/elixir/lib/enum.ex

```diff
@@ -1737,7 +1737,12 @@ defmodule Enum do
   end
 
   def reverse(enumerable, tail) do
-    reduce(enumerable, to_list(tail), fn(entry, acc) ->
+    do_reverse(enumerable, to_list(tail))
+  end
+
+  @spec do_reverse(t, list) :: list
```

Comment:
Then we should use a guard. I don't think it makes sense to add a comment
because we are worried about the performance cost of a guard.